homeassistant/components/unifiprotect/entity.py

```diff
@@ -0,0 +1,113 @@
+"""Shared Entity definition for UniFi Protect Integration."""
+from __future__ import annotations
+
+from typing import Any
+
+from pyunifiprotect.data import ProtectAdoptableDeviceModel
+
+from homeassistant.const import ATTR_ATTRIBUTION
+from homeassistant.core import callback
+import homeassistant.helpers.device_registry as dr
+from homeassistant.helpers.entity import DeviceInfo, Entity, EntityDescription
+
+from .const import DEFAULT_ATTRIBUTION, DEFAULT_BRAND, DOMAIN
+from .data import ProtectData
+
+
+class ProtectDeviceEntity(Entity):
+    """Base class for UniFi protect entities."""
+
+    _attr_should_poll = False
+
+    def __init__(
+        self,
+        data: ProtectData,
+        device: ProtectAdoptableDeviceModel | None = None,
+        description: EntityDescription | None = None,
+    ) -> None:
+        """Initialize the entity."""
+        super().__init__()
+        self.data: ProtectData = data
+
+        if device and not hasattr(self, "device"):
+            self.device: ProtectAdoptableDeviceModel = device
+
+        if description and not hasattr(self, "entity_description"):
+            self.entity_description = description
+        elif hasattr(self, "entity_description"):
+            description = self.entity_description
+
+        if description is None:
+            self._attr_unique_id = f"{self.device.id}"
+            self._attr_name = f"{self.device.name}"
+        else:
+            self._attr_unique_id = f"{self.device.id}_{description.key}"
+            name = description.name or ""
+            self._attr_name = f"{self.device.name} {name.title()}"
+
+        self._async_set_device_info()
+        self._async_update_device_from_protect()
+
+    async def async_update(self) -> None:
+        """Update the entity.
+
+        Only used by the generic entity update service.
+        """
+        await self.data.async_refresh()
+
+    @property
+    def extra_state_attributes(self) -> dict[str, Any]:
+        """Return UniFi Protect device attributes."""
+        attrs = super().extra_state_attributes or {}
+        return {
+            **attrs,
+            ATTR_ATTRIBUTION: DEFAULT_ATTRIBUTION,
+            **self._extra_state_attributes,
+        }
+
+    @callback
+    def _async_set_device_info(self) -> None:
+        self._attr_device_info = DeviceInfo(
+            name=self.device.name,
+            manufacturer=DEFAULT_BRAND,
+            model=self.device.type,
+            via_device=(DOMAIN, self.data.api.bootstrap.nvr.mac),
+            sw_version=self.device.firmware_version,
+            connections={(dr.CONNECTION_NETWORK_MAC, self.device.mac)},
+            configuration_url=self.device.protect_url,
+        )
+
+    @callback
+    def _async_update_extra_attrs_from_protect(  # pylint: disable=no-self-use
```

Comment:
Please remove the common helper for extra state attributes. We want to avoid extra state attributes. Just implement that property if really required in each platform.

Reply:
The current existing platform in this PR (camera) does not really make a lot of use of those, but the sensor/binary_sensor platforms make a lot more use of it and thus why it was broken out into its own method. The motion sensors have computed properties and `extra_state_attributes` is not supposed to have side effects. So, this runs to calculate them during the update event instead.

Reply:
Firstly we normally don't allow computed state. The only exception is times to adjust according to our entity design standards.

We'll review those platform and see if we'll approve those extra attributes. For now, we should remove this helper. Each platform can implement the extra attributes property.